tools/odps_table_tools/transform_kv_table.py

```diff
@@ -0,0 +1,318 @@
+    kv_column,
+    udf_function,
+    append_columns,
+    inter_kv_sep,
+    intra_kv_sep,
+):
+    """Generate an ODPS SQL to transform the table
+    Args:
+        input_table: input table name
+        input_table_partition: input table partition
+        output_table: output table name
+        output_columns (list): feature names
+        kv_column: kv column name
+        udf_function: udf function name
+        append_columns (list): Append column names.
+        inter_kv_sep: Inter separator in the Key-value pair string.
+        intra_kv_sep: Intra separator int Ker-value pairs string.
+    """
+    feature_names_str = ",".join(output_columns)
+    output_columns.extend(append_columns)
+    output_columns_str = ",".join(output_columns)
+    input_columns = [kv_column]
+    input_columns.extend(append_columns)
+    input_columns_str = ",".join(input_columns)
+
+    udf = """{udf}({input_col_str},
+    "{features_str}", "{inter_sep}", "{intra_sep}")
+    as ({output_col_str})""".format(
+        udf=udf_function,
+        input_col_str=input_columns_str,
+        features_str=feature_names_str,
+        output_col_str=output_columns_str,
+        inter_sep=inter_kv_sep,
+        intra_sep=intra_kv_sep,
+    )
+
+    sql = TRANSFORM_SQL_TEMPLATE.format(
+        output_table=output_table, udf=udf, input_table=input_table,
+    )
+    if input_table_partition is not None:
+        sql += " where {}".format(input_table_partition)
+    return sql
+
+
+def exec_sql(odps_entry, sql):
+    print("====> execute_sql: " + sql)
+    instance = odps_entry.run_sql(sql)
+    print("====> logview: " + instance.get_logview_address())
+    instance.wait_for_success()
+
+
+def create_udf_function(odps_entry, udf_file_path):
+    udf_resource = "sqlflow_flat_{}.py".format(int(time.time()))
+    udf_function = "sqlflow_flat_func_{}".format(int(time.time()))
+
+    delete_udf_resource(odps_entry, udf_resource)
+    resource = odps_entry.create_resource(
+        udf_resource, type="py", file_obj=open(udf_file_path)
+    )
+    print("Create python resource: {}".format(udf_resource))
+
+    delete_udf_function(odps_entry, udf_function)
+    class_type = udf_resource[0:-2] + UDF_CLASS_NAME
+    odps_entry.create_function(
+        udf_function, class_type=class_type, resources=[resource]
+    )
+
+    return udf_resource, udf_function
+
+
+def delete_udf_resource(odps_entry, udf_resource):
+    try:
+        py_resource = odps_entry.get_resource(udf_resource)
+        if py_resource:
+            py_resource.drop()
+    except odps.errors.NoSuchObject:
+        pass
+    finally:
+        print("Drop resource if exists {}".format(udf_resource))
+
+
+def delete_udf_function(odps_entry, udf_function):
+    try:
+        function = odps_entry.get_function(udf_function)
+        function.drop()
+    except odps.errors.NoSuchObject:
+        pass
+    finally:
+        print("Drop function is exists {}".format(udf_function))
+
+
+def flat_to_wide_table(
```

Comment:
Do we need make this code into a package? How do user use this code?

Reply:
The user must use those codes on MAXCOMPUTE where the can submit PyODPS jobs. There maybe two ways:
1. The user can create a[ PyODPS node ](https://help.aliyun.com/knowledge_detail/121718.html?spm=a2c4g.11186631.2.1.32192d1ddR5hFz)in DataWorks. However, the user cannot install a python package on the PyODPS node so they need to copy the UDF codes to PyODPS nodes. 
2. The user can execute by "python transform_kv_table.py --args" on the machine which can keep the connection with MAXCOMPUTE.

So, we cannot make these codes into a package because the user needs to upload the UDF codes to MAXCOMPUTE to create a UDF.